draft-ietf-jsonpath-base.md

```diff
@@ -1512,39 +1512,61 @@ The well-typedness of function expressions can now be defined in terms of this t
-   * The argument is a query (including singular query) and the declared type of the parameter is `NodesType`.
-   * The argument is a `logical-expr` and the declared type of the parameter is `LogicalType`.
-
-Note that the last bullet item includes the case that the argument is
-a query (including singular query) and the declared type of the
-parameter is `LogicalType`. In this case the nodelist resulting
-from the query is interpreted as a logical-expr in the same way
-({{extest}}) it would be converted to `LogicalType` as per {{type-conv}}.
+For a function expression to be well typed:
+
+1. its declared type must be well typed in the context in which it occurs, and
+2. its arguments must be well typed for the declared type of the corresponding parameters.
+
+(1) As per the grammar, a function expression can occur in three different
+immediate contexts, which lead to the following conditions for well-typedness:
+
+{:vspace}
+As a `test-expr` in a logical expression:
+: The function's declared result type is `LogicalType`, or
+  (giving rise to conversion as per {{type-conv}}) `NodesType`.
+
+As a `comparable` in a comparison:
+: The function's declared result type is `ValueType`.
+
+As a `function-argument` in another function expression:
+: The function's declared result type fulfills the following rules for
+  the corresponding parameter of the enclosing function.
+
+(2) The arguments of the function expression are well-typed when
+each argument of the function can be used for the declared type of the
+corresponding parameter, each according to one of the following
+conditions depending on the way the argument is supplied:
+
+{:vspace}
+As a value expressed as a literal:
+: the declared type of the parameter is `ValueType`.
+
+As a singular query:
+: the declared type of the parameter is `ValueType`.
+  In this case:
+
+  * If the query results in a nodelist consisting of a single node, the argument is the value of the node.
+  * If the query results in an empty nodelist, the argument is `Nothing`.
+
+As a general query (`filter-query`, which includes singular query):
+: the declared type of the parameter is `NodesType`.
+
+As a `logical-expr` (which includes `filter-query`):
+: the declared type of the parameter is `LogicalType`.
+  This includes the case that the argument is
+  a query (which includes singular query).
+  In this case the nodelist resulting
+  from the query is interpreted as a `LogicalType` in the same way
+  that a query in a logical context is interpreted as an existence test
+  ({{extest}}), effecting the conversion as per {{type-conv}}.
+
+As a function expression:
+: 1. the declared result type of the argument is the same as the declared
+     type of the parameter, or
+  2. the declared result type of the argument is `NodesType`, and the
+     declared type of the parameter is `LogicalType`.
+     In this case the argument is converted to `LogicalType` as per
+     {{type-conv}}.
```

Comment:
Currently this is organized by argument type then specifying what parameter type allow that argument to be valid.  The language is a bit circular.

I think it may make more sense if it's organized by parameter type then specifying what argument types are permitted.

I had restructured this section that way in a previous PR, and I felt it flowed better.

Reply:
Yes, I thought about which dimension to use as the outer and which as the inner one.
It turns out you want to use the easier to identify one (declared type) as the inner one, so that is what I chose.

Reply:
I re-read the PR with @gregsdennis's comment in mind. I didn't notice any circularity and the flow seemed good to me. Probably boils down to a matter of taste.

Reply:
Merging this now so we can get the other PRs going -- we might revisit this in another PR.

Reply:
I was just about to do the same thing, with the same rationale.